### PR TITLE
Single image implementation

### PR DESCRIPTION
**Pull request by oleksandrivantsiv**
- generic image is fully backward compatible.
- generic platform for now is default platform (make configure is not required for this platform).
- add infrastructure required to build single image.
- add single image for Mellanox platform only. "sonic-mellanox.bin" is default target.
- add all required config files to single image.
- add configuration files templates and generation of config file during startup.

### CODE REVIEW

**Comment by oleksandrivantsiv** [comment]
Single image templates implementation for Broadcom and Cavium platforms will be added as separated commits.

**Comment by lguohan** [comment]
can you check the build break?

+ sudo cp files/image_config/environment/environment ./fsroot/etc/
+ sudo cp files/image_config/environment/motd ./fsroot/etc/
+ sudo cp files/image_config/minigraph/minigraph_.xml ./fsroot/etc/sonic/minigraph.xml
cp: cannot stat 'files/image_config/minigraph/minigraph_.xml': No such file or directory

**Comment by oleksandrivantsiv** [comment]
Build failures reason is the following:
- Cavium and Broadcom platform are not read yet. Will add changes soon.
- Mellanox platform: Additional "SKU" parameter should be "make configure" command (example: make configure PLATFORM=mellanox SKU=ACS-MSN2700). Jenkins job should be changed.
- Base image job is executed under Broadcom platform:
```
[buildimage-baseimage-pr] $ /bin/bash /tmp/hudson3572960011129076848.sh
platform/broadcom/rules.mk:1: /rules.mk: No such file or directory
```
Now it can be compiled under generic platform. Generic platform is default and "make configure" is not required for it any more. To compile image the following command should be executed:
```
make target/sonic-generic.bin
```
Jenkins job should be changed.

File: files/build_templates/sonic_debian_extension.j2 [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+{% for deb in installer_debs.strip().split(' ') -%}
+sudo dpkg --extract {{deb}} $FILESYSTEM_ROOT
+{% endfor %}
+## Run depmod command for target kernel modules
+sudo LANG=C chroot $FILESYSTEM_ROOT depmod -a {{kversion}}
+{% endif %}
+{% if installer_images.strip() -%}
+{% for image in installer_images.strip().split(' ') -%}
+sudo LANG=C chroot $FILESYSTEM_ROOT docker load < {{image}}
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2]: so, we are doing this docker load in the build phase now. I thought you were thinking doing this in the boot process or onie installation process.

File: files/build_templates/sonic_debian_extension.j2 [hunk 2]
```diff
@@ -0,0 +1,115 @@
+#!/bin/bash
+
+FILESYSTEM_ROOT=$1
+[ -n "$FILESYSTEM_ROOT" ] || {
+    echo "Error: no or empty FILESYSTEM_ROOT argument"
+    exit 1
+}
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2]: can you add comment to describe what this mainly is doing?

File: files/build_templates/sonic_debian_extension.j2 [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+# Copy SNMP configuration files
+sudo cp $IMAGE_CONFIGS/snmp/snmp.yml $FILESYSTEM_ROOT/etc/sonic/
+
+# Generate system desription file
+export git_revision=$(git rev-parse --short HEAD)
+export sonic_hwsku={{sonic_hwsku}}
+export debian_version=`cat $FILESYSTEM_ROOT/etc/debian_version`
+export kernel_version={{kversion}}
+j2 $BUILD_TEMPLATES/sysDescription.j2 > sysDescription 
+sudo mv sysDescription $FILESYSTEM_ROOT/etc/ssw/
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2]: taoyu, can you take a look at this to make sure we are generate the correct sonic version here.

File: files/build_templates/docker_image_ctl.j2 [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+start() {
+    docker inspect --type container {{docker_container_name}} &>/dev/null
+    if [ "$?" -eq "0" ]; then
+        docker start -a {{docker_container_name}}
+    else
+        docker run {{docker_image_run_opt}} --name={{docker_container_name}} {{docker_image_name}}
```
Comment (lguohan) [inline files/build_templates/docker_image_ctl.j2]: do you know how long it takes to run other than the start?

**Comment by lguohan** [comment]
since docker-base is not updated, you can remove it from the changeset

File: files/build_templates/sonic_debian_extension.j2 (hunk 2 above)
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2]: maybe simply saying load docker images and configurations

---
In reply to: [98135052](https://github.com/Azure/sonic-buildimage/pull/215#discussion_r98135052) [](ancestors = 98135052)

File: files/build_templates/docker_image_ctl.j2 (hunk 4 above)
Comment (oleksandrivantsiv) [inline files/build_templates/docker_image_ctl.j2]: Approximately same time. Will do a measurement.

Comment (oleksandrivantsiv) [inline files/build_templates/docker_image_ctl.j2]: Docker run:
```
root@sonic:/home/admin# /usr/bin/swss.sh start
Run docker swss: time in miliseconds 1485530349782
Inside docker swss: time in miliseconds 1485530349961
1485530349961 - 1485530349782 = 179 miliseconds
```
Docker start:
```
root@sonic:/home/admin# /usr/bin/swss.sh start
Start docker swss: time in miliseconds 1485529942168
Inside docker swss: time in miliseconds 1485529942294
1485529942294 - 1485529942168 = 126 miliseconds
```

File: files/build_templates/sonic_debian_extension.j2 (hunk 2 above)
Comment (oleksandrivantsiv) [inline files/build_templates/sonic_debian_extension.j2]: Done

File: build_debian.sh [hunk 5]
```diff
@@ -132,15 +132,6 @@ sudo chroot $FILESYSTEM_ROOT update-initramfs -u
 ## Install latest intel igb driver
 sudo cp target/debs/igb.ko $FILESYSTEM_ROOT/lib/modules/3.16.0-4-amd64/kernel/drivers/net/ethernet/intel/igb/igb.ko
 
-## Install package without starting service
```
Comment (qiluo-msft) [inline build_debian.sh]: Why remove this block?

File: build_debian.sh [hunk 6]
```diff
@@ -268,6 +265,9 @@ sudo rm -f $ONIE_INSTALLER_PAYLOAD $FILESYSTEM_SQUASHFS
 sudo du -hs $FILESYSTEM_ROOT
 sudo mksquashfs $FILESYSTEM_ROOT $FILESYSTEM_SQUASHFS -e boot -e var/lib/docker
 
+## Compress docker files
+pushd $FILESYSTEM_ROOT && sudo tar czf $OLDPWD/$FILESYSTEM_DOCKERFS var/lib/docker; popd
+
 ## Compress together with /boot and /var/lib/docker as an installer payload zip file
-pushd $FILESYSTEM_ROOT && sudo zip $OLDPWD/$ONIE_INSTALLER_PAYLOAD -r boot/ -r var/lib/docker; popd
```
Comment (qiluo-msft) [inline build_debian.sh]: >-r var/lib/docker; [](start = 76, length = 18)

Why remove this excluding path?

File: files/image_config/ntp/ntp.conf.j2 [hunk 7]
```diff
@@ -0,0 +1,63 @@
+###############################################################################
+# Managed by Ansible
+# file: ansible/roles/acs/templates/ntp.conf.j2
```
Comment (qiluo-msft) [inline files/image_config/ntp/ntp.conf.j2]: Remove all the 'ansible' comments?

**Pull request merged**